### ISSUE TITLE
`F821 Undefined name` false negatives

### ISSUE THREAD

**Issue by nikolaik** [charliermarsh/ruff#138]
Love your work on Ruff, congrats!  ❤️

Really want to start using it soon! When initially evaluating I found a few cases where rule `F821` was not happy with me:

**Using star imports**

a.py
```python
YO = 'lo' 
```

b.py
```python
from .a import *

if YO == 'lo':  #  F821 Undefined name `YO`
    YO = 'swag'
```

**Using unpacked list values**

```python
[first] = ['yup'] #  F821 Undefined name `first`
```
   
**Inline functional declaration of typed dicts**

```python
from typing import TypedDict

class Item(TypedDict):
    nodes = list[TypedDict('Node', {'id': str})]  #  F821 Undefined name `Node`
```

**Comment by charliermarsh** [comment]
Thank you! These are great. The first and third issue were known to me but not documented anywhere (and the second is definitely a bug). Will fix.

**Comment by charliermarsh** [comment]
On star imports: right now, the best we could do for now (which is equivalent to Flake8 / PyFlakes IIUC) is provide a separate error code and message, like: `%r may be undefined, or defined from star imports: %s`. (Doing anything more advanced would break the single-file-at-a-time design of the linter, since we'd have to look at the import tree to understand what was brought in from the star import.)

**Comment by charliermarsh** [comment]
`Using unpacked list values` has been fixed in d7f95ac6b6e7775306c465a7b59e57da4e7c62d9.

**Comment by charliermarsh** [comment]
`Inline functional declaration of typed dicts` has been fixed in `6a24351202c009c14997dbb545cfb50e5393c210`.

**Comment by charliermarsh** [comment]
v0.0.31 is building now with those fixes. Going to close and create a separate issue to track starred imports. Thank you!

**Comment by nikolaik** [comment]
Thanks for all the fixes, much improved!

Testing with v0.0.31 with the fix for _Inline functional declaration of typed dicts_ in 6a24351202c009c14997dbb545cfb50e5393c210 I still see errors with:

```python
# yolo.py
from typing import TypedDict

class Item(TypedDict):
    nodes: list[TypedDict('Node', {'name': str})]
```

```console
ruff yolo.py
yolo.py:5:37: F821 Undefined name `name`

Found 1 error(s).
```

The reason why the test is passing [this line](https://github.com/charliermarsh/ruff/commit/6a24351202c009c14997dbb545cfb50e5393c210#diff-354badb80ef8370f6ea60e37eaf0074684656616254f547bedc1253fb3c81818R67), I'm guessing, is that `id` happens to be a keyword in python

**Comment by charliermarsh** [comment]
Gah, good call -- should be fixed in `c247730bf5d6394cfd4715c35ebb7d08e0da4fd0`.
